pemutil/pem.go

```diff
@@ -231,52 +231,122 @@ func ParseCertificate(pemData []byte) (*x509.Certificate, error) {
 	return nil, errors.New("error parsing certificate: no certificate found")
 }
 
-// ParseCertificateBundle extracts all the certificates in the given data.
-func ParseCertificateBundle(pemData []byte) ([]*x509.Certificate, error) {
-	var block *pem.Block
-	var certs []*x509.Certificate
-	for len(pemData) > 0 {
-		block, pemData = pem.Decode(pemData)
-		if block == nil {
-			return nil, errors.New("error decoding pem block")
+// ParseCertificateBundle returns a list of *x509.Certificate parsed from
+// the given bytes.
+//
+// - supports PEM and DER certificate formats
+//   - DER format requires WithAllowDER() option
+//   - If a DER-formatted file is given only one certificate will be returned.
+func ParseCertificateBundle(data []byte, opts ...Options) ([]*x509.Certificate, error) {
+	var err error
+	// Populate options
+	ctx := newContext("")
+	if err := ctx.apply(opts); err != nil {
+		return nil, err
+	}
+
+	filename := "input"
+	if ctx.filename != "" {
+		filename = ctx.filename
+	}
+
+	// PEM format
+	if bytes.Contains(data, PEMBlockHeader) {
+		var block *pem.Block
+		var bundle []*x509.Certificate
+		for len(data) > 0 {
+			block, data = pem.Decode(data)
+			if block == nil {
+				break
+			}
+			if block.Type != "CERTIFICATE" {
+				continue
+			}
+			var crt *x509.Certificate
+			crt, err = x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				return nil, &InvalidPEMError{
+					File: ctx.filename,
+					Err:  err,
+					Type: PEMTypeCertificate,
+				}
+			}
+			bundle = append(bundle, crt)
 		}
-		if block.Type != "CERTIFICATE" || len(block.Headers) != 0 {
-			continue
+		if len(bundle) == 0 {
+			return nil, &InvalidPEMError{
+				File:    ctx.filename,
+				Type:    PEMTypeCertificate,
+				Message: fmt.Sprintf("error parsing %s: does not contain a valid PEM encoded certificate", filename),
+			}
 		}
+		return bundle, nil
+	}
 
-		cert, err := x509.ParseCertificate(block.Bytes)
-		if err != nil {
-			return nil, errors.Wrap(err, "error parsing certificate")
+	// DER format (binary)
+	crt, err := x509.ParseCertificate(data)
+	if err != nil {
+		return nil, &InvalidPEMError{
+			File:    ctx.filename,
+			Message: fmt.Sprintf("error parsing %s as DER format: %v", filename, err),
```

Comment:
I think the errors are more verbose than necessary. 

A better approach might be to return simpler errors in the `Parse*` methods and wrap them in the `Read*` methods with the filename:
```go
return fmt.Errorf("error parsing %s: %w", filename, err)
```

Reply:
Bellow you're doing something different:
```go
return nil, &InvalidPEMError{
	File: ctx.filename,
	Type: PEMTypeCertificateRequest,
	Err:  err,
```